functorch/_src/aot_autograd.py

```diff
@@ -278,17 +279,74 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
     if config.debug_graphs:
         print("====== Forward (only) graph ======")
         fw_module.print_readable()
-    with track_graph_compiling("inference"):
-        compiled_fw = aot_config.fw_compiler(fw_module, flat_args)
+
+
+    disable_amp = should_disable_autocast(fw_module, flat_args)
+    context = disable_autocast_manager if disable_amp else nullcontext
+
+    with context():
+        with track_graph_compiling("inference"):
+            compiled_fw = aot_config.fw_compiler(fw_module, flat_args)
 
     @wraps(compiled_fw)
     def new_fn(args):
-        fw_outs = call_func_with_args(compiled_fw, args)
+        if disable_amp:
+            with disable_autocast_manager():
+                fw_outs = call_func_with_args(compiled_fw, args)
+        else:
+            fw_outs = call_func_with_args(compiled_fw, args)
         return fw_outs
 
     return new_fn
 
 
+@contextmanager
+def disable_autocast_manager():
+    guard = torch._C._DisableAutocast()
+    try:
+        yield
+    finally:
+        del guard
+
+class NonIdempotentException(Exception):
+    pass
+
+class NonIdempotentAMP(torch.fx.Interpreter):
+    """
+    When we invoke a Composite Implicit autograd operator that has an autocast rule, such as Einsum,
+    autocast is disabled during its invocation. When we trace out the operators in an implicit op,
+    re-applying autocast rules on those operators might yield divergence from what was executed at runtime.
+    This pass checks for divergence. If divergence is found, we will disable autocast.
+    We would like to avoid disabling autocast if possible because accessing TLS is slow.
```

Comment:
I'm actually not clear on this: why not just universally disable autocast during the subsequent traces?

I agree that accessing TLS can be slow in critical-path eager-mode code, but we're only setting this TLS here during aot-autograd tracing. Unless I'm missing something?

Reply:
We are also invoking it at runtime  - that's the `call_func_with_args` calls